### PR TITLE
Auth: Add skip_org_role_sync setting to OAuth integration Google

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**

Add: skip_org_role_sync for Google OAuth
- add setting for frontend
- add read of config
- add config to sample and default

**Why do we need this feature?**
Users want to have the ability to specify provider to not be synced to Grafana

**Which issue(s) does this PR fix?**:
https://github.com/grafana/grafana-enterprise/issues/4235

Fixes #

**Special notes for your reviewer**:

Demo: (most of the time is waiting to recompile the setting, so fast forward to 1.20 in the video if want to see the change

https://user-images.githubusercontent.com/7727602/212872465-21ff0fea-749b-42c6-966a-97525e581e36.mp4

### CODE REVIEW

File: pkg/login/social/social.go [hunk 1]
```diff
@@ -69,6 +69,7 @@ func ProvideService(cfg *setting.Cfg, features *featuremgmt.FeatureManager) *Soc
 	for _, name := range allOauthes {
 		sec := cfg.Raw.Section("auth." + name)
 
+		// TODO: add setting for skip_org_role_sync here
```
Comment (eleijonmarck) [inline pkg/login/social/social.go:72]:
```suggestion
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 2]
```diff
@@ -880,14 +880,28 @@ To prevent synchronization of organization roles for a specific OAuth integratio
...
 | OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
 | --- | --- | --- | --- |
 | AzureAD | false | false | will sync with AzureAD roles |
 | AzureAD | true | false | skip org role sync for OAuth providers including AzureAD users |
 | AzureAD | false | true | skip org role sync for AzureAD users |
 | AzureAD | true | true | skip org role sync for AzureAD users and all other OAuth providers |
 
+### [auth.google] skip_org_role_sync
+
+Opon first login from a user we set the organization roles from the setting `AutoAssignOrgRole`. If you want to manage organizational roles, you can set the `skip_org_role_sync` option to `true`. Please note that there is also a separate setting called `oauth_skip_org_role_update_sync` which has a different scope. While `skip_org_role_sync` only applies to the specific OAuth provider, `oauth_skip_org_role_update_sync` is a generic setting that affects all configured OAuth providers.
```
Comment (eleijonmarck) [inline docs/sources/setup-grafana/configure-grafana/_index.md:893]: As we do not get roles from Google auth provider.

Here is the setting that we set eventually after the code goes through a user that logs in without a role assigned from the provider.
https://github.com/grafana/grafana/blob/main/pkg/services/user/userimpl/user.go#L151

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 3]
```diff
@@ -68,3 +68,13 @@ When a user logs in using an OAuth provider, Grafana verifies that the access to
 Grafana uses a refresh token to obtain a new access token without requiring the user to log in again. If a refresh token doesn't exist, Grafana logs the user out of the system after the access token has expired.
 
 By default, Grafana includes the `access_type=offline` parameter in the authorization request to request a refresh token.
+
+## Skip organization role sync
+
+We do not currently sync roles from Google and instead set the AutoAssigned role to the user at first login. To be able to manage your users organization role from within Grafana , set `skip_org_role_sync` to `true`.
```
Comment (eleijonmarck) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:74]: @kalleep should we let them know that we are not syncronizating the roles. Than we should not mention it in the docs. 
I think we should mention it, as it is the truth and they could read it from the code anyways. 🤔 

Wdyt?

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 4]
```diff
@@ -880,14 +880,28 @@ To prevent synchronization of organization roles for a specific OAuth integratio
...
+
+The setting `oauth_skip_org_role_update_sync` will be deprecated in favor of provider-specific settings.
+
+The table below show the OAuth provider and their setting with the default value and the skip org role sync setting.
+| OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
+| --- | --- | --- | --- |
+| Google | false | false | user organization roles are set with defaultRole and will not be able to be changed |
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:900]:
```suggestion
| Google | false | false | User organization roles are set with `defaultRole` and cannot be changed |
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 5]
```diff
@@ -880,14 +880,28 @@ To prevent synchronization of organization roles for a specific OAuth integratio
...
+| Google | true | false | user organization roles are set with defaultRole for Google, for other providers the synchronization will be skipped and the org role will be able to be changed; along with other OAuth provider users org roles |
+| Google | false | true | user organization roles are set with defaultRole and the org role will be able to be changed for Google synced users |
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:902]:
```suggestion
| Google | false | true | User organization roles are set with `defaultRole` and the org role can be changed for Google synced users. |
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 6]
```diff
@@ -880,14 +880,28 @@ To prevent synchronization of organization roles for a specific OAuth integratio
...
+| Google | true | true | user organization roles are set with defaultRole for Google, for other providers the synchronization will be skipped and the org role will be able to be changed; along with other OAuth provider users org roles |
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:903]:
```suggestion
| Google | true | true | User organization roles are set with `defaultRole` for Google. For other providers, the synchronization will be skipped, and the org role can be changed, along with other OAuth provider users' org roles. |
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md (hunk 3 above)
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:74]:
```suggestion
We do not currently sync roles from Google and instead set the AutoAssigned role to the user at first login. To manage your user's organization role from within Grafana, set `skip_org_role_sync` to `true`.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 2 above)
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:893]:
```suggestion
Upon the first login from a user, we set the organization roles from the setting `AutoAssignOrgRole`. If you want to manage organizational roles, set the `skip_org_role_sync` option to `true`. 

> **Note:** There is a separate setting called `oauth_skip_org_role_update_sync` which has a different scope. While `skip_org_role_sync` only applies to the specific OAuth provider, `oauth_skip_org_role_update_sync` is a generic setting that affects all configured OAuth providers.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 7]
```diff
@@ -880,14 +880,28 @@ To prevent synchronization of organization roles for a specific OAuth integratio
...
 | AzureAD | false | true | skip org role sync for AzureAD users |
 | AzureAD | true | true | skip org role sync for AzureAD users and all other OAuth providers |
 
+### [auth.google] skip_org_role_sync
+
+Opon first login from a user we set the organization roles from the setting `AutoAssignOrgRole`. If you want to manage organizational roles, you can set the `skip_org_role_sync` option to `true`. Please note that there is also a separate setting called `oauth_skip_org_role_update_sync` which has a different scope. While `skip_org_role_sync` only applies to the specific OAuth provider, `oauth_skip_org_role_update_sync` is a generic setting that affects all configured OAuth providers.
+
+The setting `oauth_skip_org_role_update_sync` will be deprecated in favor of provider-specific settings.
+
+The table below show the OAuth provider and their setting with the default value and the skip org role sync setting.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:897]:
```suggestion
The following table shows the OAuth provider's setting with the default value and the skip org role sync setting.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 8]
```diff
@@ -880,14 +880,28 @@ To prevent synchronization of organization roles for a specific OAuth integratio
 
 The setting `oauth_skip_org_role_update_sync` will be deprecated in favor of provider-specific settings.
 
-The following table shows the OAuth providers, the default value setting, and the skip org role sync setting.
+The table below show the OAuth provider and their setting with the default value and the skip org role sync setting.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:883]:
```suggestion
The following table shows the OAuth provider's setting with the default value and the skip org role sync setting.
```

**Review by Jguer — APPROVED** [review]

**Review by chri2547 — APPROVED** [review]
Docs LGTM! Thank you.

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 9]
```diff
@@ -880,14 +880,30 @@ To prevent synchronization of organization roles for a specific OAuth integratio
...
 | AzureAD | false | true | skip org role sync for AzureAD users |
 | AzureAD | true | true | skip org role sync for AzureAD users and all other OAuth providers |
 
+### [auth.google] skip_org_role_sync
+
+Upon the first login from a user, we set the organization roles from the setting `AutoAssignOrgRole`. If you want to manage organizational roles, set the `skip_org_role_sync` option to `true`. 
+
+> **Note:** There is a separate setting called `oauth_skip_org_role_update_sync` which has a different scope. While `skip_org_role_sync` only applies to the specific OAuth provider, `oauth_skip_org_role_update_sync` is a generic setting that affects all configured OAuth providers.
+
+The setting `oauth_skip_org_role_update_sync` will be deprecated in favor of provider-specific settings.
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:897]: nit: should this be a part of the above note?

**Review by IevaVasiljeva — COMMENTED** [review]
Do I understand correctly that for Google this change impacts only the frontend, as there is no role syncing mechanism on the backend anyway?

**Comment by eleijonmarck** [comment]
@IevaVasiljeva 
> Do I understand correctly that for Google this change impacts only the frontend, as there is no role syncing mechanism on the backend anyway?

Yea that's correct. We do not get roles from Google currently
https://github.com/grafana/grafana/blob/main/pkg/login/social/google_oauth.go#L19

This is currently not implemented

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 9 above)
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:897]: yea, it is a bit verbose currently. I want to review the docs at some point soon, to either make it more concise or to put all the tables together of sorts.

I will remove this line

**Pull request merged**